<a href="https://colab.research.google.com/github/smahasr/Codsot----Machine-Learning-Intern-Task-3/blob/main/Machine_learning_intern_task_3_Codsoft.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Title :CUSTOMER CHURN PREDICTION
## Objective :
Develop a model to predict customer churn for a subscription based service or business. Use historical customer data, including features like usage behavior and customer demographics, and try algorithms like Logistic Regression, Random Forests, or GradientBoosting to predict churn.

Step 1 : Load dataset and preprocessing

In [ ]:
import pandas as pd
import numpy as np

df=pd.read_csv('/content/Churn_Modelling.csv')
df.head()

RowNumber  CustomerId   Surname  CreditScore Geography  Gender  Age  \
0          1    15634602  Hargrave          619    France  Female   42   
1          2    15647311      Hill          608     Spain  Female   41   
2          3    15619304      Onio          502    France  Female   42   
3          4    15701354      Boni          699    France  Female   39   
4          5    15737888  Mitchell          850     Spain  Female   43   

   Tenure    Balance  NumOfProducts  HasCrCard  IsActiveMember  \
0       2       0.00              1          1               1   
1       1   83807.86              1          0               1   
2       8  159660.80              3          1               0   
3       1       0.00              2          0               0   
4       2  125510.82              1          1               1   

   EstimatedSalary  Exited  
0        101348.88       1  
1        112542.58       0  
2        113931.57       1  
3         93826.63       0  
4         79084.10       0

In [ ]:
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler, LabelEncoder

numeric_columns = df.select_dtypes(include=np.number).columns
df[numeric_columns] = df[numeric_columns].fillna(df[numeric_columns].mean())

label_encoder_geo = LabelEncoder()
label_encoder_gender = LabelEncoder()
df['Geography'] = label_encoder_geo.fit_transform(df['Geography'])
df['Gender'] = label_encoder_gender.fit_transform(df['Gender'])


features = ['CreditScore', 'Geography', 'Gender', 'Age', 'Tenure', 'Balance',
            'NumOfProducts', 'HasCrCard', 'IsActiveMember', 'EstimatedSalary']
target = 'Exited'
X = df[features]
y = df[target]


scaler = StandardScaler()
X[['CreditScore', 'Age', 'Tenure', 'Balance', 'NumOfProducts', 'EstimatedSalary']] = scaler.fit_transform(
    X[['CreditScore', 'Age', 'Tenure', 'Balance', 'NumOfProducts', 'EstimatedSalary']]
)


<ipython-input-2-22cf65438011>:21: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  X[['CreditScore', 'Age', 'Tenure', 'Balance', 'NumOfProducts', 'EstimatedSalary']] = scaler.fit_transform(


Step 2: Model Selection and model Evaluation

In [ ]:
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier, GradientBoostingClassifier
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

log_reg = LogisticRegression()
rf = RandomForestClassifier()
gb = GradientBoostingClassifier()

log_reg.fit(X_train, y_train)
rf.fit(X_train, y_train)
gb.fit(X_train, y_train)

log_reg_preds = log_reg.predict(X_test)
rf_preds = rf.predict(X_test)
gb_preds = gb.predict(X_test)


Step 3: Model Evaluation

In [ ]:
def evaluate_model(predictions, y_test):
    accuracy = accuracy_score(y_test, predictions)
    precision = precision_score(y_test, predictions)
    recall = recall_score(y_test, predictions)
    f1 = f1_score(y_test, predictions)
    return accuracy, precision, recall, f1

log_reg_metrics = evaluate_model(log_reg_preds, y_test)
print(f'Logistic Regression - Accuracy: {log_reg_metrics[0]}, Precision: {log_reg_metrics[1]}, Recall: {log_reg_metrics[2]}, F1 Score: {log_reg_metrics[3]}')

rf_metrics = evaluate_model(rf_preds, y_test)
print(f'Random Forest - Accuracy: {rf_metrics[0]}, Precision: {rf_metrics[1]}, Recall: {rf_metrics[2]}, F1 Score: {rf_metrics[3]}')

gb_metrics = evaluate_model(gb_preds, y_test)
print(f'Gradient Boosting - Accuracy: {gb_metrics[0]}, Precision: {gb_metrics[1]}, Recall: {gb_metrics[2]}, F1 Score: {gb_metrics[3]}')


Logistic Regression - Accuracy: 0.8155, Precision: 0.6, Recall: 0.183206106870229, F1 Score: 0.2807017543859649
Random Forest - Accuracy: 0.869, Precision: 0.7763713080168776, Recall: 0.4681933842239186, F1 Score: 0.584126984126984
Gradient Boosting - Accuracy: 0.8655, Precision: 0.7540983606557377, Recall: 0.4681933842239186, F1 Score: 0.5777080062794349


Step 4: Outofsample prediction

In [ ]:
best_model = gb

new_data = [[650, 1, 0, 35, 5, 50000, 2, 1, 1, 60000]]
new_data_scaled = scaler.transform(new_data)
churn_prediction = best_model.predict(new_data_scaled)
print(f'Churn Prediction: {churn_prediction}')

Churn Prediction: [1]


/usr/local/lib/python3.10/dist-packages/sklearn/base.py:439: UserWarning: X does not have valid feature names, but StandardScaler was fitted with feature names
  warnings.warn(
/usr/local/lib/python3.10/dist-packages/sklearn/base.py:439: UserWarning: X does not have valid feature names, but GradientBoostingClassifier was fitted with feature names
  warnings.warn(


Conclusion:

The Random Forest model achieved the highest performance with an accuracy of 86.9%, precision of 77.6%, recall of 46.8%, and F1 score of 58.4%  and followed by Gradient boost and logistic regression. The out-of-sample prediction for the new data indicates the customer is predicted to churn (prediction: 1).